src/Chainweb/Pact/Backend/ChainwebPactDb.hs

```diff
@@ -0,0 +1,482 @@
+
+createVersionedTablesTable  :: BlockHandler SQLiteEnv ()
+createVersionedTablesTable =
+  callDb $ \db -> exec_ db
+    "CREATE TABLE VersionedTables (tablename TEXT\
+  \ , blockheight UNSIGNED BIGINT\
+  \ , version UNSIGNED BIGINT\
+  \ , CONSTRAINT versionTableConstraint UNIQUE(version, blockheight));"
+
+createVersionedTable :: TableName -> BlockVersion -> BlockHandler SQLiteEnv ()
+createVersionedTable name b = do
+  callDb $ \db -> exec_ db $
+                  "CREATE TABLE "
+                   <> Utf8 (toS (flip T.snoc ']' $ T.cons '[' $ asString name))
+                   <> " (rowkey TEXT,blockheight UNSIGNED BIGINT, version UNSIGNED BIGINT,txid UNSIGNED BIGINT,tabledata BLOB)"
+                   -- rename tabledata to rowdata or just data
+  versionedTablesInsert name b
+
+type ParentHash = BlockHash
+
+handleVersion :: BlockHeight -> Maybe ParentHash -> BlockHandler SQLiteEnv ()
+handleVersion _ Nothing = return ()
+handleVersion bRestore (Just hsh) = do
+  bCurrent <- do
+    r <- callDb $ \ db -> qry_ db "SELECT max(blockheight) AS current_block_height FROM BlockHistory;" [RInt]
+    SInt bh <- liftIO $ expectSingleRowCol "detectVersionChange (block):" r
+    return $ BlockHeight (fromIntegral bh)
+  vCurrent <- do
+    r <- callDb $ \db -> qry_ db "SELECT max(version) AS current_version FROM VersionHistory;" [RInt]
+    SInt version <- liftIO $ expectSingleRowCol "detectVersionChange (version):" r
+    return $ ReorgVersion version
+
+  -- REWRITE ERROR WITH INTERNALERROR PREFIX
+  -- enforce invariant that the history has (B_restore-1,H_parent).
+  historyInvariant <- callDb $ \db -> do
+        res <- qry db "SELECT COUNT(*)\
+           \ FROM BlockHistory\
+           \ WHERE blockheight = ?\
+           \ AND hash = ?;"
+            [SInt $ fromIntegral $ pred bRestore, SBlob (Data.Serialize.encode hsh)] [RInt]
+            >>= expectSingleRowCol "detectVersionChange (historyInvariant):"
+        return res
+
+  when (historyInvariant /= SInt 1) $
+    internalError "History invariant violation"
+
+  case compare bRestore (bCurrent + 1) of
+   GT -> internalError "PactInternalError: detectVersionChange: Block_Restore invariant violation!"
+   EQ -> assign bsBlockVersion (BlockVersion bRestore vCurrent)
+   LT -> do
+     let bvEnv = (BlockVersion bRestore (succ vCurrent))
+     assign bsBlockVersion bvEnv
+     tableMaintenanceRowsVersionedTables bvEnv
+     dropVersionedTables bvEnv
+     deleteHistory bvEnv
+
+tableMaintenanceRowsVersionedTables :: BlockVersion -> BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedTables bv = do
+  tableMaintenanceRowsVersionedSystemTables bv
+  tableMaintenanceRowsVersionedUserTables bv
+
+tableMaintenanceRowsVersionedSystemTables :: BlockVersion -> BlockHandler SQLiteEnv ()
+tableMaintenanceRowsVersionedSystemTables (BlockVersion bh _) = callDb $ \db -> do
+  exec' db "DELETE FROM [SYS:KeySets] WHERE blockheight >= ?" [SInt (fromIntegral bh)]
```

Comment:
Should this be `>` instead of `>=`? You're `restore`ing to `bh` right, not to one before it?

Reply:
Please review the algorithm